### PR TITLE
README and Templates: Update files based on new guides/templates

### PR DESCRIPTION
**Pull request by thatblindgeye**
#### This commit

Related to https://github.com/TheOdinProject/curriculum/pull/23587

I updated the PR and Issue templates to be more consistent across our repos.

I also updated the README file to make it more focused on this repo's focus, while still linking to the curriculum repo.

#### Checklist
 - [x] You have read [The Odin Projects Contributing Guide](https://github.com/TheOdinProject/theodinproject/wiki/Contributing-Guide)?
 - [ ] You have verified the tests and linters all pass against your changes?
 - [ ] You have included automated tests for your changes where applicable?

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/bug_report.md
Comment (KevinMulhern) [inline .github/ISSUE_TEMPLATE/bug_report.md:16]: Can we keep the example please? it helps prime the bug report author with what format they should provide.

File: .github/ISSUE_TEMPLATE/feature_request.md
Comment (KevinMulhern) [inline .github/ISSUE_TEMPLATE/feature_request.md:17]: Same here please

**Review by KevinMulhern — CHANGES_REQUESTED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (thatblindgeye) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: So to give context for this change, it's mainly to have more consistent templates across repos where possible (aside from obvious deviations, like the curriculum repo we don't need users to include automated tests). Basically so users can open a PR without major differences when filling out the template/it's instantly recognizable how it should be filled out.

The reason for using "Description of the Changes" was basically to be a catch-all for both the "why" of the PR and the "how", since it neve really seemed like having those two separated was all that necessary. I can see what you mean by it being *too* broad, though. Beyond that, I don't think I really have any strong opinion either way on format (other than whatever the format,  I think if we can it should be something that can work across a majority of our repos with tweaks mainly to the checkboxes).

What would you think of something along the lines of the following tweaks to the original format:

```
1. Reason for PR
<!-- 
If your PR closes an open issue, link it by including the `Closes` keyword followed by the issue number/issue URL. For example, `Closes #2013`.

Otherwise, please explain the reason for your pull request, e.g. what problem it solves or what benefit it provides.
 -->

2. Outline of Changes
<!-- 
A bullet point list of one or more items outlining what was done in this PR to solve the problem or introduce the feature/enhancement.
 -->
```
Suggesting these tweaks just because if the worry is that users will just mention what's already in the PR, the previous headings on their own didn't really convey what that section should actually include all that well. At least, not without reading the commented out instructions (which the same could be said for the "Description of Changes" instructions below this line).

This would remove the existing "Related issue" step and place it in the first "reason" step, and I think this could be a template that could be used in most of our other repos (which is a big plus in my book).

File: .github/ISSUE_TEMPLATE/bug_report.md
Comment (thatblindgeye) [reply .github/ISSUE_TEMPLATE/bug_report.md:16]: Did you want the example *just* for the "how to reproduce" section, or also for the "expected behavior" section?

File: .github/ISSUE_TEMPLATE/feature_request.md
Comment (thatblindgeye) [reply .github/ISSUE_TEMPLATE/feature_request.md:17]: So I'm not exactly sure if this examples really makes sense in this section as proposed, just because it's more "here's why I think this feature would be a good idea". That said, I could just add back in the Acceptance Criteria section and keep the example (or a similar example) there. Let me know what you think.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (KevinMulhern) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: This summarises the approach I use for commit messages: https://thoughtbot.com/blog/5-useful-tips-for-a-better-commit-message

Splitting the why and the how has a bunch of benefits:

* It's easier on the author since the prompts provide clarity with what they need to write which helps keep the message to the point.
* It's easier on the reviewer to get context at a glance as to why the changes in the PR are necessary and how the author resolved the issues. 
* It saves people time in the future. For example, if someone comes across a line of code that they don't understand, or perhaps they just want to know the context as to why it was added in the first place. They can git blame and find out without having to parse the important details from potentially multiple paragraphs of text that open-ended prompts often have.

I've been through a lot of git commit message conventions before landing on this one a few years ago. It's been the best I've come across so far. I'd go so far as to recommend we use it on all repos for the reasons listed above.

I'm not totally against updating the "Because" and "This commit" naming conventions. _But_ looking back at the last few PR's from contributors outside of the maintainer team. It seems like they aren't having issues with filling them in correctly.

**Review by KevinMulhern — CHANGES_REQUESTED** [review]

File: .github/ISSUE_TEMPLATE/feature_request.md
Comment (thatblindgeye) [reply .github/ISSUE_TEMPLATE/feature_request.md:17]: Yeah, I think I removed it to make things more "general", but thinking about it again I think this section can really work across any repo (and users can just skip it if they don't have any criteria, while maintainers can edit in any criteria as well)

File: .github/ISSUE_TEMPLATE/feature_request.md
Comment (thatblindgeye) [reply .github/ISSUE_TEMPLATE/feature_request.md:6]: This info we need as it generates info on the actual issue selection screen as well as applies default labels/title to the issue.

So each time a user submits a "feature request" issue, it'll automatically have the "Status: Needs Review" label applied. Because the intent is for any issues that haven't been assigned yet to require reviewing from a maintainer to see whether the issue is valid/can be assigned, having this label auto-added just relieves some tedious upkeep that I and a couple others have been doing.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (thatblindgeye) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: Definitely understand those benefits. I had been used to the other repos that never were setup with this format, so coming to this format was more jarring for me than vice versa 😆

Like I said, ultimately I don't think I have strong opinions either way re: because/this commit vs reason for/outline of, or any other variation. Ultimately what I'd prefer is a format that, outside of repo-specific checkboxes, is more consistent and easily recognizable regardless of the repo for any experience level of user (obviously we can't capture *every* user, though).

If you believe the "because/this commit" format coudl hit that nail,, I'd be down for implementing this sort of format across other repos as well. Otherwise, if you feel this format is more beneficial to this repo only, I'd also be down to just having this main repo use one PR template and the other repos use another (because this repo is much more different than the other ones in terms of content).

File: .github/ISSUE_TEMPLATE/feature_request.md
Comment (thatblindgeye) [reply .github/ISSUE_TEMPLATE/feature_request.md:22]: To be clear, would this dummy text be for an Acceptance Criteria section, or the Description of the Feature Request section?

Comment (KevinMulhern) [reply .github/ISSUE_TEMPLATE/feature_request.md:22]: Just the description please mate

File: .github/ISSUE_TEMPLATE/feature_request.md
Comment (KevinMulhern) [reply .github/ISSUE_TEMPLATE/feature_request.md:6]: ahh gotcha, TIL. Really neat 😄

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (KevinMulhern) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: Any repos with a codebase can definitely benefit from it like this one and the exercise repos. I think the curriculum can as well. Most of the PR's we get are tweaks to the existing content so having structured explanations for changes made could be immensely helpful.

I've been using this format on the curriculum repo 😆
https://github.com/TheOdinProject/curriculum/pull/23212
https://github.com/TheOdinProject/curriculum/pull/23219

Comment (thatblindgeye) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: So the last thing I think to clarify (don't quote me on this later 😆 ), is the issue/PR title format. This PR uses a "location of change: description" format, whereas I've noticed you use more of a "keyword: description" format.

The reason I had opted for the "location" of the change is to let contributors know immediatley where an issue/PR pertains to, which along with the description should help avoid duplicates being made. So using your two examples above with this format, it would have ended up being "Foundations: Archive old Frontend lessons" and "Intermediate HTML section: Add Form Basics lessons".

It may not need to be an either/or situation, as maybe both the keyword and location could work together (e.g. `Feature - Intermediate HTML: Add Form Basics lesson`), but let me know what you think.

Comment (KevinMulhern) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: I _really_ like the `location of change: description` format for the curriculum, I think that will be a great system. It makes a lot of sense and I can definitely see its utility in being able to spot duplicated issues.

But I don't think it would work as well when changes will be mostly code related as they will be in this and the exercise repos. Just because the location of change would mostly pertain to files and there may be many files changed from different parts of the app in any given PR.

I'd prefer we use scrum story types in this repo please:
* Feature - adds new or amends user-facing behaviour
* Chore - changes that have no user-facing value, refactors, dependency bumps etc.
* Fix - Bug fixes.

Comment (thatblindgeye) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: I definitely wasn't gonna follow up with clarification on what you view a "chore" as, as I wasn't sure if the chore label description is all that accurate 😆 

Sounds good, that difference definitely makes sense. I'll update things and re-request your review. Thanks for all the input!

Comment (KevinMulhern) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: Sounds good, thanks for putting this all together @thatblindgeye 💪

**Comment by thatblindgeye** [comment]
@KevinMulhern Made updates per our discussions. Some notes:

- For the Description of Feature Request section, I wasn't entirely sure what to use as dummy text, nor whether utilizing the `When X occurs, I want Y to happen, to achieve Z` would be significantly more beneficial than a description of the suggested feature. If you could provide an example using a recent feature on here that'd be great, otherwise I've kept this as-is for now.
- I tweaked the Acceptance Criteria section for Feature Request so that checkboxes could be used. My thought was as someone works on an issue, they would check each checkbox and the issue could be considered completed only after all checkboxes are checked. Not entirely sure if this then makes a Bug Report less easy to track without checkboxes, but let me know what you think.
- For the PR Template, I put in the heading verbiage I had previously mentioned so you could see how it looked as a whole. No strong opinions, though, so I can change it to "Because" and "This PR" instead (only because "This Commit" doesn't sound entirely accurate if multiple commits are made on a PR).
- Again for the PR Template, though this time would like your opinion: how do you feel about the checkboxes that must be completed for re-requesting reviews and ensuring issues are linked?

**Comment by KevinMulhern** [comment]
Hey @thatblindgeye, sorry for the delay.

#### Description of Feature Request section
It would be beneficial for the same reasons as the commit message prompts I reckon. It gives the author a standardised structured template they can use instead of a free form description. There would be less to think about which would ultimately make the feature request easier to write. 

I haven't used them with TOP yet, only at work. But some examples would be like this:

> When I'm on the paths page, then I want the paths badge to show my progress, so I can easily track my progress on that path.

> When I'm on my dashboard, then I want to be able to resume where I left off on a course, so I can get back to making progress quicker.

> When I start TOP again after a long break, then I want this to be able to reset my progress, so I can start afresh.

#### Acceptance Criteria
You can think of the acceptance criteria as a workflow the user would follow for the desired outcome for the feature and each item as a step in that workflow. That's why I had them as a numbered list.

For example: 
1. Log in
2. Visit a lesson page
3. Mark it as complete

Checkboxes will probably work fine too though, happy to keep that change. However, I don't think it would work for bug reports since the steps to reproduce and expected behaviour is more useful as a numbered list the developer can easily follow.

#### PR Template Verbiage
Happy to concede and try out the new wording 😄 


####  PR Template Checkboxes
I think it would be worth keeping the ensuring issues are linked checkbox. But the re-requesting reviews may be redundant.

**Comment by thatblindgeye** [comment]
No worries! I'd rather this have time taken so that once this is all resolved, we should hopefully not really have to touch these templates again unless there's a significant issue with the format or someone suggests a *really* good idea.

I believe I addressed your most recent comments in the PR and/or in my comments below; let me know if I missed anything.

**Description of Feature Request section:** I added some verbiage in more as a "following this format can help" thing rather than a "this must be filled out exactly". I _think_ there maybe more info needed as to why X feature request would be beneficial or what problem it would solve, which the "When X, I want Y, so that Z" format may not always be able to convey on its own.

**Acceptance Criteria:** Agreed on the bug report. For the Acceptance Criteria itself, my imagining was more having it not formatted as a list of steps to follow like the expected behavior of a bug, but rather "these items must be completed for this feature/enhancement to be considered resolved". I think this could be something that just works universally, e.g.:

```
- [ ] A reset button resets a users progress (for main repo)
- [ ] The resource linked in the description above is added as an Additional Resource (for curriculum repo)
- [ ] The exercise detailed in the description above is added to the repo (for one of our exercise repos)
```

**PR Template:** Reverted the verbiage with the only tweak being "This PR" rather than "This Commit"; like I said I didn't feel strongly 😆 I also have grown to like the flow of reading a PR: "This PR wants to implement a new feature to add a dark mode toggle to the site. Because XYZ. This is a quick explanation of how this PR achieves that."

I did end up removing both the "request another review" and "linked an issue" checkboxes, though, based on some convo that occurred on the server with other maintainers. Having that `Closes #XXXXX` text baked into the template I would agree with Sully in that it seems most have done a good job filling that out when needed since it was implemented. It being the first step now rather than the last step I think will further help.

**Review by KevinMulhern — APPROVED** [review]
Looks great to me @thatblindgeye. Thanks for all the the time and effort you've put into this! lets get it shipped 🚢

**Pull request merged**
